### ISSUE TITLE
static curl with static nghttp2

### ISSUE THREAD

**Issue by svnpenn** [curl/curl#2516]
the current situation with curl and nghttp2 is... not good. Given this file:

http://github.com/curl/curl/blob/master/docs/examples/simple.c

You can try to compile like this:

~~~
x86_64-w64-mingw32-gcc -static -Wl,--start-group -lbcrypt -lcrypt32 -lcrypto \
-lcurl -lgdi32 -lnghttp2 -lssh2 -lssl -lwldap32 -lws2_32 -lz simple.c
~~~

Result:

~~~
simple.c:(.text+0xe8): undefined reference to `__imp_curl_easy_init'
~~~

You can fix that with `-DCURL_STATICLIB`, but then another problem follows:

~~~
libcurl.a(libcurl_la-http2.o): (.text+0x29): undefined reference to
`__imp_nghttp2_session_get_stream_user_data'
~~~

You can try adding `-DNGHTTP2_STATICLIB`, but the error will not go away. It
seems the problem can only be resolved by using `-DNGHTTP2_STATICLIB`
**at compile time of libcurl**.

This is bad. If a user has a `libcurl.a` file, it should be safe to assume that
file will only require other static libraries. Some suggestions:

1. If a user builds static libcurl, then `gcc -static` should be all thats
   needed to build a static program with libcurl. users should need to use
   `-static` and `-DCURL_STATICLIB` when building a static program with libcurl
   
2. if user builds static libcurl, static nghttp2 should be assumed, further
   `-DNGHTTP2_STATICLIB` shouldnt be required at any step of building curl or
   building a program with libcurl

**Comment by svnpenn** [comment]
@jay i think the problem is deeper than you are suggesting

i would prefer not to build libcurl, as the "static" library is already available via Cygwin and MSYS2

however in both cases the "static" library (libcurl.a) is requiring a shared nghttp2. while you could argue that both the cygwin and msys2 maintainers built the static libcurl "wrong", it shouldnt be this hard. that is to say that the invocation you have suggested shouldnt be needed to build a static libcurl that has only static requirements

**Comment by jay** [comment]
They didn't build it wrong. If you want to prefer static dependencies over dynamic then you must specify that. And in some cases you may then have to figure out the dependencies of those dependencies, and so on. I don't think it is our responsibility.

**Comment by svnpenn** [comment]
> They didn't build it wrong.

yeah, they did. they created a static library that depends on a shared library.
that is an error. i dont blame them though, i blame whoever wrote the part of
the curl codebase that allows for a situation where someone could build a
static libcurl that relied on shared libraries.

what is the use case for that? maybe im crazy but you either build everything
shared or everything static. for example, my FFmpeg build pulls in libfdk-aac
and libx264, but when i build FFmpeg i do this:

    configure --enable-libfdk-aac --enable-libx264 --extra-ldflags=-static

not this:

    configure --enable-libfdk-aac --enable-libx264 \
    --extra-ldflags='-static -DFDKAAC_STATICLIB -DX264_STATICLIB'

because it doesnt make sense to need those other arguments: if i say `-static`
that means everything static, all the way down. also, why only
`-DNGHTTP2_STATICLIB`? Curl pulls in all these:

* -lbcrypt
* -lcrypt32
* -lcrypto
* -lgdi32
* -lnghttp2
* -lssh2
* -lssl
* -lwldap32
* -lws2_32
* -lz

so why is nghttp2 special?

**Comment by bagder** [comment]
The `NGHTTP2_STATICLIB` define for a static nghttp2 build is mandated by the nghttp2 library (in the exact same vein we tell users to define `CURL_STATICLIB`) so that's not something we can control.

Windows is just a finicky system which makes it hard to build both a static and a shared lib at the same time (which is what we do on *nix-style systems).

The reliance on these defines is an unfortunate build step, but we still decided to do this many years ago since the alternative approaches caused us even more grief. :frowning_face:

**Comment by svnpenn** [comment]
> Windows is just a finicky system which makes it hard to build both a static
> and a shared lib at the same time (which is what we do on \*nix-style
> systems).

@bagder dont you mean, impossible? with the current system:

1. if you **dont define** `NGHTTP2_STATICLIB` when building libcurl, then a
   shared libcurl is built, and static libcurl is built **that relies on shared
   nghttp2**

2. if you **do define** `NGHTTP2_STATICLIB` when building libcurl, then a
   shared libcurl is built **that relies on static nghttp**, and static libcurl
   is built

**Comment by bagder** [comment]
I meant when doing libraries in general, whatever library.

Yes, when you (on Windows) want to build a static libcurl that builds with a static nghttp2, you need to compile both libcurl to know about it and your application to know about it.

**Comment by svnpenn** [comment]
@bagder perhaps you could explain why its needed to use `-DCURL_STATICLIB`
and `-DNGHTTP2_STATICLIB` in order to create a purely static libcurl

it seems like we have a case of "it is that way, because it is that way". maybe
i am not understanding, but i havent seen a good reason yet **why** curl is the
way it is in regard to static library.

with perhaps every other library/program i have ever built, you dont need to do
this. you simply pass `-static` somewhere in the process and thats it

why is curl so cumbersome in this regard? what is the benefit?

**Comment by bagder** [comment]
> it seems like we have a case of "it is that way, because it is that way".

I'm describing what we have and how it works. It's not the same as *it has to be this way for the rest of time*.

The `CURL_STATICLIB` logic was added in commit 3ccbed10 back in 2004 to make us not have to use a .def file when building DLLs on Windows. That was a win because we kept breaking it.
